### PR TITLE
Add set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "-Wl,--no-as-needed") to CMakeLists.txt

### PR DESCRIPTION
**Pull request by pbelevich**
This is a fix for a rare build issue on Ubuntu:
`symbol lookup error: miniconda3/envs/pytorch-py3.7/lib/libmkl_intel_lp64.so: undefined symbol: mkl_blas_dsyrk`
https://software.intel.com/en-us/articles/symbol-lookup-error-when-linking-intel-mkl-with-gcc-on-ubuntu

### CODE REVIEW

**Comment by ezyang** [comment]
cc @soumith if there are binary implications, but this looks reasonable. Is there any way to reasonably test that we've done this correctly?

**Review by ezyang — APPROVED** [review]
There should be a comment explaining what goes wrong without this, esp. because there is relevant back story here.

**Comment by ezyang** [comment]
cc @xuhdev

File: CMakeLists.txt [hunk 1]
```diff
@@ -192,6 +192,14 @@ cmake_dependent_option(
 
 set(ONNX_NAMESPACE "onnx_torch" CACHE STRING "A namespace for ONNX; needed to build with other frameworks that share ONNX.")
 
+# This is a fix for a rare build issue on Ubuntu:
+# symbol lookup error: miniconda3/envs/pytorch-py3.7/lib/libmkl_intel_lp64.so: undefined symbol: mkl_blas_dsyrk
+# https://software.intel.com/en-us/articles/symbol-lookup-error-when-linking-intel-mkl-with-gcc-on-ubuntu
+if(NOT APPLE)
+  set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "-Wl,--no-as-needed")
+  set(CMAKE_SHARED_LINKER_FLAGS_DEBUG "-Wl,--no-as-needed")
```
Comment (xuhdev) [inline CMakeLists.txt]: Adding `--no-as-needed` looks reasonable to me. My comments:

- One single line of `set(CMAKE_SHARED_LINKER_FLAGS "-Wl,--no-as-needed")` would be effective on all build types.
- What would happen on Windows?

**Comment by soumith** [comment]
wouldn't this break for the stub libs that we compile (for example _thnvrtc.so)? if it passes CI, then go for it.

File: CMakeLists.txt (hunk 1 above)
Comment (pbelevich) [reply CMakeLists.txt]: > One single line of set(CMAKE_SHARED_LINKER_FLAGS "-Wl,--no-as-needed") would be effective on all build types.
Addressed.

> What would happen on Windows?
I think it's a platform specific issue and platform specific fix(ubuntu/mkl/ld). Windows builds passed, but I restricted if condition to `if(UNIX AND NOT APPLE)`, see below

> Don't think it hurts to have this line on MacOS, and therefore no need of if(NOT APPLE).
It does: `ld: unknown option: --no-as-needed`

**Comment by pbelevich** [comment]
@soumith how to test stub libs?

File: CMakeLists.txt [hunk 2]
```diff
@@ -192,6 +192,13 @@ cmake_dependent_option(
 
 set(ONNX_NAMESPACE "onnx_torch" CACHE STRING "A namespace for ONNX; needed to build with other frameworks that share ONNX.")
 
+# This is a fix for a rare build issue on Ubuntu:
+# symbol lookup error: miniconda3/envs/pytorch-py3.7/lib/libmkl_intel_lp64.so: undefined symbol: mkl_blas_dsyrk
+# https://software.intel.com/en-us/articles/symbol-lookup-error-when-linking-intel-mkl-with-gcc-on-ubuntu
+if(UNIX AND NOT APPLE)
```
Comment (xuhdev) [inline CMakeLists.txt]: Is `NOT APPLE` necessary here? The same error may occur if some user on MacOS has customized default compiler configuration to something similar to the default of Ubuntu.

Comment (pbelevich) [reply CMakeLists.txt]: It fails on pytorch_macos_10_13_cuda9_2_cudnn7_py3_build: ld: unknown option: --no-as-needed

Comment (xuhdev) [reply CMakeLists.txt]: I see, makes sense, macOS's default linker does not support this option. You can go for `if(LINUX)` if it feels better, as I don't think other BSD systems (if we would ever support them)  ship a linker that supports this option.

**Review by xuhdev — APPROVED** [review]

**Comment by pbelevich** [comment]
Unfortunately faced #21700, needs further investigation

**Comment by pbelevich** [comment]
Alright, figured out that it was an issue with LLD 8.0.0, ready for review

**Comment by xuhdev** [comment]
> Alright, figured out that it was an issue with LLD 8.0.0, ready for review

Did you make any new changes?

**Comment by pbelevich** [comment]
> Did you make any new changes?

I had to switch back from LLD 8.0.0 to GNU ld version 2.27-34.base.el7, no changes in this PR

**Pull request closed without merging**